src/service/compact/retention.rs

```diff
@@ -59,27 +102,64 @@ pub async fn delete_by_stream(
         lifecycle_end
     );
 
-    // Hack for 1970-01-01
-    if lifecycle_start.le("1970-01-01") {
-        let lifecycle_end = created_at + Duration::try_days(1).unwrap();
-        let lifecycle_end = lifecycle_end.format("%Y-%m-%d").to_string();
-        return db::compact::retention::delete_stream(
+    // Create tasks with break on red days
+    let lifecycle_start = format!("{}T00:00:00Z", lifecycle_start);
+    let lifecycle_end = format!("{}T00:00:00Z", lifecycle_end);
+    let start = DateTime::parse_from_rfc3339(&lifecycle_start)?.with_timezone(&Utc);
+    let end = DateTime::parse_from_rfc3339(&lifecycle_end)?.with_timezone(&Utc);
+    if start.ge(&end) {
+        return Ok(0); // created_at is after lifecycle_end, just skip
+    }
+
+    let original_time_range = TimeRange::new(
+        start.timestamp_nanos_opt().expect("valid timestamp") / 1000,
+        end.timestamp_nanos_opt().expect("valid timestamp") / 1000,
+    );
+
+    let mut final_time_ranges = vec![];
+    red_days.iter().for_each(|red_day| {
+        let _ranges_added = populate_time_ranges_for_deletion(
+            &mut final_time_ranges,
+            red_day,
+            &original_time_range,
+        );
+    });
+
+    // if red days is empty, then just delete the whole time range
+    if red_days.is_empty() {
+        final_time_ranges.push(original_time_range);
+    }
+
+    let job_nos = final_time_ranges.len();
+
+    for time_range in final_time_ranges {
+        let time_range_start = Utc
+            .timestamp_nanos(time_range.start * 1000)
+            .format("%Y-%m-%d")
+            .to_string();
+        let time_range_end = Utc
+            .timestamp_nanos(time_range.end * 1000)
+            .format("%Y-%m-%d")
+            .to_string();
+        log::debug!(
+            "[COMPACT] delete_by_stream {}/{}/{}/{},{}",
             org_id,
             stream_type,
             stream_name,
-            Some((lifecycle_start, lifecycle_end.as_str())),
+            time_range_start,
+            time_range_end,
+        );
+
+        db::compact::retention::delete_stream(
+            org_id,
+            stream_type,
+            stream_name,
+            Some((time_range_start.as_str(), time_range_end.as_str())),
         )
-        .await;
+        .await?;
     }
 
-    // delete files
-    db::compact::retention::delete_stream(
-        org_id,
-        stream_type,
-        stream_name,
-        Some((lifecycle_start, lifecycle_end)),
-    )
-    .await
+    Ok(job_nos as u32)
 }
```

Comment:
In this function there are some many time and timestamp converting, maybe we can use timestamp for comparing.